proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+## Scenarios and User Experience
+
+We would expect that users would be able to write C# code that can do simple Swift interop. Additionally, we would expect that for cases where the interop system does not support seamless interop, developers could write a shim in Swift that could be called from C# code. Developers should not need to write a shim in C or Assembly to interact with Swift APIs.
+
+## Requirements
+
+### Goals
+
+In short, we should completely eliminate the required C/assembly sandwich that's currently required to call Swift from C# code, and potentially vice versa. In particular, neither C# nor Swift users should have to deal with lower-level system state, like registers or stack state, to call Swift from C#.
+
+### Non-Goals
+
+C# and Swift are different languages with different language semantics. It is not a goal to map every construct from one language to the other. However, there are some terms in both languages that are sufficiently similar that they can be mapped to an identical semantic term in the other language. Interop should be seen as a Venn diagram where each language forms its own circle, and interop is in the (much smaller) space of equivalent terms that are shared between them.
+
+## Stakeholders and Reviewers
+
+- [@dotnet/interop-contrib](https://github.com/orgs/dotnet/teams/interop-contrib)
+- [@dotnet/macios](https://github.com/orgs/dotnet/teams/macios)
+- [@lambdageek](https://github.com/lambdageek)
+- [@SamMonoRT](https://github.com/SamMonoRT)
+- [@kotlarmilos](https://github.com/kotlarmilos)
+- [@JulieLeeMSFT](https://github.com/JulieLeeMSFT)
+- [@amanasifkhalid](https://github.com/amanasifkhalid)
+- [@davidortinau](https://github.com/davidortinau)
+- [@jkotas](https://github.com/jkotas)
+
+## Design
+
+### Runtime
+
+#### Calling Conventions
+
+Swift has two calling conventions at its core, the "Swift" calling convention, and the "SwiftAsync" calling convention. We will begin by focusing on the "Swift" calling convention, as it is the most common. The Swift calling convention has a few elements that we need to contend with. Primarily, it allows passing arguments in more registers. Additionally, it has two dedicated registers for specific components of the calling convention, the self register and the error register. The runtime support for this calling convention must support all of these features. The additional registers per argument is relatively easy to support, as each of our compilers must already have some support for this concept to run on Linux or Apple platforms today. We have a few options for the remaining two features we need to support.
```

Comment:
Are these two calling conventions or just an optional register usage when involving "async"? Not trying to quibble or terms, but if it is simply a register being used when async is involved, then I wouldn't phrase this as two and instead call out that as another thing to consider like the self and error semantics below.

Reply:
Technically, there are more calling conventions. In addition to "Swift" and "SwiftAsync",  there is `enable-library-evolution` which uses a similar calling convention in order to manage value type layouts that may change. In addition, you can annotate any function or function type with `@convention (foo)` where today foo is a member of the set {c, block, objc}. Finally, if a class inherits from `NSObject` and has the `@objc` attribute, the compiler will writer an objective C dispatcher for every method. This dispatcher ends up calling the vtable methods.

Reply:
@stephen-hawley Technically, `enable-library-evolution` is not a separate calling convention, but it does change the ABI.

Here's my definition of "calling convention" for this section (I'll add it do the doc): "A distinct calling convention is represented by a specific calling convention name in LLVM IR".

The Swift compiler doesn't use a separate calling convention for Library Evolution. Instead it changes how the arguments are passed to the call (changing parameters from their struct types to pointers) while still using the Swift or SwiftAsync calling convention.

I'll make sure to mention the `@convention` option and `@objc` as well.